Loadless Timer and Death Counter

eddynya opened the pull request:
**Death Counter**

- Keeps track of your deaths in story mode

**Loadless Timers**

- Includes both fullgame and segment timers, and each can be toggled to not show at all, show at all times, show between worlds, and only show at the end of the run
- If the segment timer is enabled in any way, at the end of the run, you can see all your split times + segment times for each world
- Toggle-able warning that shows up on the name entry screen if no timers are enabled 

**Unified `draw_timer()` Code**

- The challenge mode segment timer, IW timer, RTA/Pause timer, and story timer all use the updated function 

**Known Issues** 

- everything is explained here https://docs.google.com/document/d/1b2wigdLeCfmix8JAe3KXI8Z4aBpLXtcXYd08oKKzuYw/edit?usp=sharing

eddynya committed:
loadless timer + segment timer stuff

eddynya committed:
fixing segment timer after resetting

eddynya committed:
somewhat fixed some issues with the timer missing frames

eddynya committed:
fixed the missing frames issue (kinda jank fix) for the loadless timer

eddynya committed:
fixed segments not showing up when run ends

eddynya committed:
added a story death counter

eddynya committed:
fixed timer/death counter spacing + timer not on warning

eddynya committed:
some code cleanup

eddynya committed:
unified timer code, fixed spacing on old timers

eddynya committed:
make the loadless timer count during timeover

eddynya committed:
cut out parts of the timer

-timer no longer increments on mkb::STAGE_SELECTED due to (load?) variance
-timer no longer increments between the stage select a press and the start of the 10 ball screen spin in bc of variance between emu and console (will test more though to be sure)

eddynya committed:
adjusted timer + fixed death counter bug

-the timer now increments for exactly 49 frames during fade out (which is the usual fade out time when stage selecting) after pressing stage select. this fade out time should not vary between sd/usb/dolphin
-fixed the death counter bug where it would increment to 1 on the first stage automatically

eddynya committed:
some cleanup

ComplexPlane commented on scripts/gen_compile_commands.py:
You can undo these permissions changes with `git checkout scripts` inside WSL. I think this is caused by having the repo cloned outside WSL, or accessing the repo inside WSL with a tool outside WSL... I forget what.

Point is, IMO the best experience is to `git clone` the repo from a WSL terminal inside your WSL home directory, and open it in vscode using `code .` inside the folder in a WSL terminal. I don't know what your setup looks like but nambo/rehtrop should definitely be able to sort you out if there's an issue.

rehtropsmb replied:
Setting up my environment on my laptop and I ran into this too. I believe when vscode isn't connected to WSL, it attempts to Windows-ify the files? I'm also not sure if `code .` sets up the remote connection if you don't already have the WSL extension installed.

@eddynya If you don't have the WSL extension for vscode, you'll want to get it before trying Complex's solution. Once you have it, do ctrl+shift+p and type "connect to wsl" and select that option. Then try running `git checkout scripts` and the files should be reverted.

If this doesn't work and you want a simpler solution you can try `git config core.filemode false`. This will stop git from picking up these file changes.

ComplexPlane replied:
I'm not sure about `git config core.filemode false` because we do want to respect executable bit of the script files and future scripts etc. You shouldn't run into this issue if you clone the repo inside your WSL home dir, and open code with `code .` (or Connect to WSL).

ComplexPlane commented on src/mods/storytimer.cppZone.Identifier:
Remember to delete this (again I think it's caused by inside/outside WSL shenanigans)

ComplexPlane commented on src/mods/storytimer.cpp:
When you get a chance you should format your code:

1. Install the C++ extension as well as the clangd extension
2. Run "format code" from the command palate or whatever the Windows keyboard shortcut is

I like using these settings to make it automatic:
- Auto save on focus change
- Format on save

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+
+namespace storytimer {
+
+    enum class FullgameTimerOptions {
+        F_DontShow = 0,
+        F_AlwaysShow = 1,
+        F_BetweenWorlds = 2,
+        F_EndOfRun = 3,
+    };
+    
+    enum class SegmentTimerOptions {
+        S_DontShow = 0,
+        S_AlwaysShow = 1,
+        S_BetweenWorlds = 2,
+        S_EndOfRun = 3,
+    };
```
`enum class` options shouldn't have a prefix like `F_` because using the enum class name like `FullgameTimerOptions::` separates the scope from that of other enum classes. This is in contrast to a plain `enum` which should have prefixed options, because all options of all `enum`s in the same scope need to have a unique name. That's why they have funky prefixes in `mkb2_ghidra.h`.

Also, I think these two enums should be a single `TimerOption` enum since they're identical (and probably used in the same way)?

eddynya replied:
Sure, I'll remove the prefixes. Having two separate enums does have use cases I think, some of them being:

- If I want to run timer blind but still want to see how good my individual world segments are, having them separated like this accomplishes that (so I would set the fullgame timer to show at the end of the run, and the segment timer to show at all times or between worlds for this)
- Maybe instead I don't like how 2 lines of text looks like, so I want the fullgame timer always enabled, but I still want to see how I did on each world. So for that I could have the segment timer show between worlds only

ComplexPlane replied:
You can accomplish that with one enum type but two variables, both with the same type

ComplexPlane commented on src/mods/deathcounter.cpp:
```diff
@@ -0,0 +1,63 @@
+    // to do: however, if you retry after breaking the tape on the very first frame (so the frame before goal init), it does count as a death when it should not
+    if (mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT || mkb::g_storymode_stageselect_state == mkb::STAGE_SELECT_INTRO_SEQUENCE) {
+        s_can_die = false;
+    }
```
Why should retrying on first frame not count as a death? Wouldn't it be simpler to keep it consistent with the game's behavior?

ComplexPlane commented on src/mods/deathcounter.cpp:
```diff
@@ -0,0 +1,63 @@
+
+static bool s_can_die;
+static u32 s_death_count;
+static bool s_in_story;
```
One important piece of overall feedback I have for your PR is: try to minimize the number of static variables you use. In general, the more state that needs to be kept track of between frames, the harder it is to reason about what the code will do since the "state space" is larger aka there's more possibilities.

Some signs that something shouldn't be a static variable:

* It is (or easily _could be_) computed on each frame
* It's only used by one function like `tick()` or `disp()`
    * Or if it's used by multiple, they could easily compute it themselves by e.g. calling a shared function to do it
* You don't care what the value on the previous frame was

To illustrate it with this file, it definitely makes sense for `s_death_count` to be a static variable b/c how else would you track deaths across frames, but I don't think `s_can_die` and `s_in_story` need to be. Those bools look like they're only used in `tick()` and `disp()` respectively, and seem like they can be computed on each frame.

ComplexPlane commented on src/systems/menu_defn.cpp:
```diff
@@ -658,6 +658,51 @@ static Widget s_cm_seg_widgets[] = {
+    "Always show",
+    "Between worlds",
+    "End of run",
+};
```
Like the `enum class` these can probably be a single array

ComplexPlane commented on src/systems/pref.cpp:
```diff
@@ -342,8 +358,8 @@ static DefaultU8Pref s_default_u8_prefs[] = {
-    u8 bool_prefs[8];  // up to 64 bool prefs
-    u8 u8_prefs[25];   // 25 u8 prefs
+    u8 bool_prefs[9];  // up to 64 bool prefs
+    u8 u8_prefs[27];   // 27 u8 prefs
```
The bool prefs comment is wrong now but can you actually just delete these comments? It's annoying to update them

ComplexPlane commented on src/utils/timerdisp.cpp:
```diff
@@ -9,33 +9,88 @@ static constexpr u32 MINUTE_FRAMES = SECOND_FRAMES * 60;
 static constexpr u32 HOUR_FRAMES = MINUTE_FRAMES * 60;
 
 static constexpr s32 X = 380;
+static constexpr s32 X_2 = 18;
```
Unused

ComplexPlane commented on src/mods/deathcounter.cpp:
```diff
@@ -0,0 +1,63 @@
+    if (s_can_die == true && (mkb::sub_mode == mkb::SMD_GAME_READY_INIT || mkb::sub_mode == mkb::SMD_GAME_RINGOUT_INIT 
+    || mkb::sub_mode == mkb::SMD_GAME_TIMEOVER_INIT || mkb::sub_mode == mkb:: SMD_GAME_SCENARIO_RETURN)){
+        // you can die either by retrying after dropping in, falling out, timing over, or stage selecting after dropping in (but before breaking the tape)
+        s_death_count += 1;
```
Is doing this without checking if the game is paused a potential bug? Does it make sense to check if the game is paused w/ an early return at the start of `tick()`?

rehtropsmb replied:
`s_can_die` would be set to false so it seems like this would only be able to be ran once anyways

ComplexPlane commented on src/mods/deathcounter.cpp:
```diff
@@ -0,0 +1,63 @@
+            return;
+        }
+   draw::debug_text(18, 56, draw::WHITE, "Deaths: "); 
+   draw::debug_text(98, 56, draw::WHITE, "%d", s_death_count);
```
Can you use some multiple of `draw::DEBUG_CHAR_WIDTH` for the X value for well-defined spacing?

ComplexPlane commented on src/mods/deathcounter.cpp:
```diff
@@ -0,0 +1,63 @@
+
+void disp() {
+
+    if (s_in_story == false || freecam::should_hide_hud() || pref::get(pref::BoolPref::ShowDeathCounter) == false ){
```
You don't need to use `== true` or `== false` to check boolean variables/expressions in `if` statements. The thing inside `()` is a boolean expression that computes a boolean value. So for example, to check if a variable is true, just do `if (my_var) { ... }`, and for false, `if (!my_var) { ... }`.

ComplexPlane commented on src/mods/deathcounter.cpp:
```diff
@@ -0,0 +1,63 @@
+   draw::debug_text(18, 56, draw::WHITE, "Deaths: "); 
+   draw::debug_text(98, 56, draw::WHITE, "%d", s_death_count);
+
+   // to do so I don't forget: get pref for the storytimers so that the the text doesn't overlap
```
Reminder

eddynya replied:
Forgot to delete that comment, I ended up adding code to `storytimer.cpp` that handled what row to draw the timers on depending on if the death counter was enabled

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+static constexpr s32 segment_timer_text_offset = 44;
+static constexpr s32 IW_time_location_x = 42+24; 
+static constexpr s32 IW_time_text_offset = 32;
+static constexpr s32 stage_fade_out_time = 49;
```
Constants should be `ALL_UPPER_CASE`

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+static u32 s_segment_timer[11]; // IW timer for world k
+static u32 s_split[11];    // s_split[k] is the loadless time on tape break of the 10th stage of world k
+static u32 s_segment_start_time[11];  // the loadless time at the start of world k, used to calculate s_segment_timer[k]
+static bool s_can_change_segment_start_time[11];
```
Pretty sure all of these arrays should have a size of 10, and the `k` you use to index them should start at 0

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+static bool s_is_timeover;
+static bool s_can_increment_stage_counter;
+static bool s_lower_stage_counter;
+static bool s_start_stage_fade_out_timer;
```
Most of these (variables that aren't timers) look like they can be local variables?

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+static u32 s_spin_in_timer_correction;
+static u32 s_game_scenario_return_timer_correction;
+static u32 s_world_start_timer_correction;
+static u32 s_loadless_story_timer;
```
It might be cleaner to stick your timer variables into a struct like `TimerGroup`. That way it's easy set them all to zero by setting the struct to `{}`, and you don't have to suffix every variable with `_timer`

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+        }
+
+    // code for handling loadless split and segment times
+    for (s32 k=1; k<11; k++){
```
Should probably start `k` at 0

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+    
+    // for later use, it's useful to record how many stages we've completed
+    // there is likely a simpler way to increment the stage counter in a way that works when you retry after breaking the tape and also does not break if you pause during 
+    // game goal init or game goal main
```
If it simplifies things, you might be able to use `mkb::get_world_unbeaten_stage_count(world_idx)` and `mkb::scen_info.world` (current world)

eddynya replied:
I chose to rewrite the existing code without using 'mkb::get_world_unbeaten_stage_count' because that increments after the fade to white happens and not on tape break. My code for completed stages does increment on tape break though.  I'm not sure if my rewrite is simpler, or if using 'mkb::get_world_unbeaten_stage_count' and when I need something involving the number of completed stages to happen on tape break to just check if we are in GAME_GOAL_INIT and if we are not paused. I can try out the 2nd option if you think it would simplify the code further

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+
+   if (mkb::sub_mode == mkb::SMD_GAME_GOAL_MAIN) {
+        // not allowed to increment the stage counter on goal init because of first framing
+        s_can_increment_stage_counter = true;
```
should this be set to false otherwise?

eddynya replied:
I simplified/rewrote the existing completed stages code (see the comment above)

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+        s_is_on_spin_in = true;
+    } else {
+        s_is_on_spin_in = false;
+    }
```
For stuff like this you can set the bool var to the result of the stuff inside parens of the `if` statement instead of using an `if`

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+
+    // story mode states entered during the story select screen
+    if (mkb::g_storymode_stageselect_state == mkb::STAGE_SELECT_INTRO_SEQUENCE || mkb::g_storymode_stageselect_state == 3 || 
+        mkb::g_storymode_stageselect_state == mkb::STAGE_SELECT_IDLE || mkb::g_storymode_stageselect_state == 5 || mkb::g_storymode_stageselect_state == mkb::STAGE_SELECTED) {
```
There's a `StoryModeStageSelectState` enum you can make use of, but it looks like it's missing `5`

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+        }
+    }
+
+}
```
I hope that reading the current world and # of beaten stages in the current world like I described before could simplify this... I'm have a pretty hard time parsing it

I'm hoping if you know what world/stage you're on, you simply increment the appropriate per-world timers and don't need a lot of this logic?

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+                s_display_story_timer = true;
+            } else {
+                s_display_story_timer = false;
+            }
```
Can simplify this to `s_display_story_timer = s_is_between_worlds` (and it should be a local variable)

ComplexPlane commented on src/utils/timerdisp.cpp:
Ok, at this point I actually think `timerdisp` is trying to be way too many things for its own good. I did suggest generalizing `draw_timer()`, but I didn't imagine you'd want to do stuff like optionally draw two timers on the same line in a specific format.

Here's the relevant concerns as I understand them:

1. Multiple modules want to draw timers left-aligned at certain parts of the screen, but they should "stack" to avoid overlapping
2. Modules need to format their timers differently (prefixes, more than one timer, etc.), but without duplicating excessive formatting code

Modules need to negotiate _where_ their lines of text are rendered, but aside from that, they don't need to agree on how their timers are drawn.

So here's my proposal:

1. Make a separate module that is dedicated to dishing out "slots" (`(x, y)` coordinate pairs) for other modules to render arbitrary lines of text to. Modules that call this function earlier get "slots" higher on the screen. The function might look like:

```c++
enum class Region {
    TopLeft,
    TopRight
};
Vec2d get_next_slot(Region pos);
```

Although, since your code is the only thing using the top left region, you might not need to implement this at all / just keep the current logic for how stuff in the top-right region is drawn.

2. Let individual modules render their own arbitrary line of text. Modules could use a shared function to format a single frame counter variable as a time as a helper, and e.g. call it multiple times if they need to render more than one timer on the same line.

Btw, this is how you build a formatted string by itself, so you can use it later:

```c++
char str[32] = {}; // Must be big enough to hold the longest possible string, plus one null character
mkb::sprintf(buf, "int: %02d, float: %.02f, string: %s", 10, 1.5, "substring");
```

Using a shared timer formatting function would look something like:

```c++
char timer_str[32] = {};
timer_utils::format_timer(timer_str, my_framecount /* More options here */);
draw::debug_text("SEG: %s", timer_str);
```

Note that you can't `return`a pointer to the buffer you `sprintf` into because it is a stack variable, so you have to pass it "down" instead of "up".

rehtropsmb replied:
All the timer code here looks like it's gotten overly complex, so this sounds like a great idea! @eddynya I can help you get started with this, just dm me if you have questions!

ComplexPlane reviewed (changes requested):
This should be plenty to get you started!

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+            +s_spin_in_timer_correction+s_game_scenario_return_timer_correction+ s_world_start_timer_correction;
+
+
+        if (s_is_on_spin_in == true) {
+            // increment the timer every frame during spin in
+            s_spin_in_timer +=1;
+        }
+        if (s_is_on_gameplay == true) {
+            //increment the timer every frame during gameplay
+            s_gameplay_timer +=1;
+        }
+        if (s_is_postgoal == true && s_stage_fade_out_timer <= stage_fade_out_time) {
+            // increment the timer every frame after game goal init happens; once you press stage select, a separate 49 frame timer is started (fade out from stage select
+            // to the first completely white frame takes 49 frames). once the timer hits 49 frames, stop incrementing the timer 
+            // until the 10 ball screen starts spinning in
+            s_postgoal_timer +=1;
+        }
+        if (mkb::g_storymode_stageselect_state == mkb::STAGE_SELECT_INTRO_SEQUENCE || mkb::g_storymode_stageselect_state == 3 || 
+        mkb::g_storymode_stageselect_state == mkb::STAGE_SELECT_IDLE) {
+            //increment the timer every frame on the story mode select screen until the a press input; we do not include the transition time after pressing a afterwards 
+            // even ignoring completely white frames, the time spent on mkb::g_storymode_stageselect_state == mkb::STAGE_SELECTED can be
+            // highly variable (up to over 40 frames sometimes!), so for the purpose of a loadless timer, it makes sense to cut this out from the timer
+            s_stage_select_timer +=1;
+        }
+        if (s_is_on_exit_game_screen == true) {
+            // increment the timer every frame on the exit game screen
+            s_exit_game_timer +=1;
+        }
+        if (s_is_on_fallout_screen == true) {
+            // increment the timer every frame during the fallout sequence and y/n screen
+            s_fallout_timer += 1;
+        }
+        if (s_is_timeover == true) {
+            // increment the timer every frame during the timeover sequence
+            s_timeover_timer += 1;
+        }
```
You probably want to use `else` for a lot of these...

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,427 @@
+    // if the segment timer is enabled in any capacity, show all 10 split times + iw times after the tape is broken on the last stage
+    if (SegmentTimerOptions(pref::get(pref::U8Pref::SegmentTimerOptions)) != SegmentTimerOptions::S_DontShow) {
+       if (s_is_run_complete == true) {
+            // I'm so sorry :(
+            // I don't know how to get the text to show "Wk" where k ranges in a for loop
```
You can use `mkb::sprintf` like I talked about in another comment

ComplexPlane commented:
Oh just noticed the conflicts... yeah you will need to pull and rebase/merge master to reconcile w/ rehtrop's latest changes

eddynya committed:
small fixes before tackling bigger issues

-now using 1 enum class for both timers instead of 2

eddynya committed:
Delete storytimer.cppZone.Identifier

ComplexPlane commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,490 @@
+static u32 s_game_scenario_return_timer_correction;
+static u32 s_world_start_timer_correction;
+static u32 s_loadless_story_timer;
+static bool s_in_story;
+static bool s_is_on_world[11];
+static bool s_is_between_worlds;
+static bool s_is_run_complete;
+static bool s_is_on_spin_in;
+static bool s_is_on_gameplay;
+static bool s_is_postgoal;
+static bool s_is_on_stage_select_screen;
+static bool s_is_on_exit_game_screen;
+static bool s_is_on_fallout_screen;
+static bool s_is_timeover;
+static bool s_can_increment_stage_counter;
+static bool s_lower_stage_counter;
+static bool s_start_STAGE_FADE_OUT_TIMEr;
```
Vast majority of these don't need to be static variables (or perhaps not even variables at all)

eddynya replied:
I made most of these local variables, if you think most don't need to be variables at all I can do that. My reasoning for keeping them as variables is that (I hope) writing it this way makes it easier to read than having like 20 submodes in an if()

eddynya committed:
small changes

-rewrote completed stages code (will probably try and simplify still)
-small changes to death counter

Commit:
Merge branch 'master' of https://github.com/rehtropsmb/SMB2PracticeMod

Resolving merge conflicts with rehtrop's changes

-Changed the RTA/Pause timer to use the new function I made as a bandaid fix until I address Complex's comment
-Changed one of the timer widgets to say "Segment & Loadless Timers" until I decide how I want to reorganize this part of the menu

Commit:
Merge branch 'master' of github.com:ComplexPlane/SMB2PracticeMod
Pulled updated ghidra headers

eddynya committed:
updated names to match new ghidra header

eddynya committed:
starting to rewrite segment timer code

(not finished)

eddynya committed:
another approach

eddynya committed:
fixed some issues with the rewrite

Commit:
pre cleanup backup

eddynya committed:
clean up arc 2 part 1

eddynya committed:
added count deaths on stage 1 toggle to deathcounter

eddynya committed:
optimized memory used by timer

removed some static u32 variables that were being kept track of

rehtropsmb commented on src/systems/pref.cpp:
```diff
@@ -370,8 +391,8 @@ static DefaultU8Pref s_default_u8_prefs[] = {
 struct PrefState {
-    u8 bool_prefs[8];  // up to 64 bool prefs
-    u8 u8_prefs[26];   // 26 u8 prefs
+    u8 bool_prefs[10];
```
You don't need to increase the number of bool prefs here quite yet. Each array value can hold up to 8 bool prefs, so an array length of 8 can hold up to 64 prefs. Changing this to 10 means it can hold 80 prefs which isn't necessary.

rehtropsmb commented on src/systems/menu_defn.cpp:
```diff
@@ -1259,6 +1321,21 @@ static Widget s_displays_widgets[] = {
+                .pref = pref::BoolPref::ShowDeathCounter,
+            },
+    },
+    */
```
Is this still needed or can it be removed?

eddynya replied:
I just forgot to delete this, thanks for reminding me.

rehtropsmb commented on src/mods/deathcounter.cpp:
```diff
@@ -0,0 +1,70 @@
+    if (mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT ||
+        mkb::g_storymode_stageselect_state == mkb::STAGE_SELECT_INTRO_SEQUENCE) {
+        s_can_die = false;
+    }
```
You may be able to solve this in a similar way to how I fixed IL Mark validation by tracking when the ball travels thru the goal, I can help you with that when my PR is merged :)

rehtropsmb commented on src/mods/deathcounter.cpp:
```diff
@@ -0,0 +1,70 @@
+        if (mkb::sub_mode == mkb::SMD_GAME_PLAY_MAIN) {
+            s_can_die = true;
+        }
+    }
```
Could rewrite as

```
    // Don't increment the death counter on stage 1 if the setting is ticked
    if (mkb::sub_mode == mkb::SMD_GAME_PLAY_MAIN) {
        if (pref::get(pref::BoolPref::CountStage1Deaths) {
            s_can_die = true
        } else if (!pref::get(pref::BoolPref::CountStage1Deaths && storytimer::get_completed_stagecount() != 0) {
            s_can_die = true
        }
    }
```

rehtropsmb commented on src/systems/pref.h:
```diff
@@ -56,6 +56,9 @@ enum class BoolPref : u8 {
     ReverseMode,
+    StoryTimerWarning,
+    ShowDeathCounter,
+    CountStage1Deaths,
```
Rename to `CountFirstStageDeaths`?

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+                              s_timer_group[9].segment, true, false, draw::WHITE);
+    }
+
+    // show warning on the name entry screen if no timers are on (if the toggle for the warning is
+    // turned on)
+    if (pref::get(pref::BoolPref::StoryTimerWarning) &&
+        TimerOptions(pref::get(pref::U8Pref::FullgameTimerOptions)) == TimerOptions::DontShow &&
+        TimerOptions(pref::get(pref::U8Pref::SegmentTimerOptions)) == TimerOptions::DontShow &&
+        mkb::scen_info.mode == 21) {
+        // mkb::scen_info.mode 21 is the name entry screen, not sure if it has a name in ghidra
+        draw::debug_text(460, 425, draw::RED, "Timer Not On!");
+    }
```
What's the use case for this? Wouldn't you see the timer appear when you start your run? I think it might be better to have the reverse scenario where it says something like "Timer Enabled" when you're on the name entry screen and you have the timer on. (then you won't need an extra pref)

eddynya replied:
I just went ahead and deleted the story timer warning entirely

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+    EndOfRun = 3,
+};
+
+static u32 s_loadless_story_timer;
```
Is `s_loadless_story_timer` only used to draw the full timer on screen? I think you may be able to calculate this value by adding the splits from the timer group, which would make this unnecessary.

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+        s_completed_stages = 0;
+        s_start_stage_fade_out_timer = false;
+        s_stage_fade_out_timer = 0;
+        for (s32 k = 0; k < 10; k++) {
```
Use the constant mentioned above, `WORLD_COUNT`, in your loops, `for (u32 k = 0; k < WORLD_COUNT; k++) {`

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+    u32 non_gameplay;
+    u32 world_start_correction;
+};
+static TimerGroup s_timer_group[10];  // each world has its own TimerGroup structure
```
You should add a variable `static constexpr u32 WORLD_COUNT = 10;` and replace all instances of `10` with that constant since it seems like you use it a lot. Maybe also add a seperate `STAGES_PER_WORLD` constant too? (even though they both have the same value of 10)

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+struct TimerGroup {
+    u32 segment;     // the time taken to complete a world up until tape break on the last stage
+    u32 full_world;  // the time taken to complete a world until the fade to white on the last stage
+    u32 split;       // the time from the start of the run until tape break of the last stage of the
```
Split seems like it could be calculated each frame by adding `segment`'s and `full_world`'s. I think you should try to get rid of this value if it isn't absolutely needed.

eddynya replied:
done

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+    u32 split;       // the time from the start of the run until tape break of the last stage of the
+                     // world
+    u32 gameplay;
+    u32 non_gameplay;
```
Can one of these be calculated each frame by doing some math? Like could you calculate `non_gameplay` by doing `non_gameplay = full_world - gameplay`? I think one of these values isn't necessary.

eddynya replied:
I removed the `non_gameplay` and `gameplay` variables from TimerGroup, so I'll resolve all comments relating to that

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+                     // world
+    u32 gameplay;
+    u32 non_gameplay;
+    u32 world_start_correction;
```
Not too sure what this is, could you give an explanation? To me, it doesn't seem like something that needs to be stored in a static variable either.

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+static bool s_start_stage_fade_out_timer;
+static u32 s_stage_fade_out_timer;
+static s32 s_completed_stages;            // the completed stages for the whole run
+static s32 s_completed_stages_world[10];  // the completed stages in world k
```
Do you need both of these? You should be able to calculate the completed stages in a world by using modulo of 10. Example: if 53 stages are completed, `57 % 10` returns 7. If you need to get the number of worlds that have been cleared, you can divide. `57 / 10` returns 5. This would clear up the 10 static variables in the `s_completed_stages_world` array.

I could see why you'd want to have the number of completed stages on a per-world basis if this timer were meant to be used for any%, or if there was a pack with less or more than 10 stages in each world. I don't think either of these situations are too important though (because most people wanting a loadless timer are probably not doing any% and a custom pack with varied worldcounts doesn't exist yet).

eddynya replied:
removed the `s_completed_stages_world` array

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+        for (s32 k = 0; k < 10; k++) {
+            s_timer_group[k] = {};
+            s_completed_stages_world[k] = 0;
+        }
```
Your code could be easier to read if you break out some of the chunks of logic into seperate functions. For example, this section could be taken out into a seperate `reset_timer` function.

```
static void reset_timer() {
    s_can_lower_stage_counter = false;
    s_loadless_story_timer = 0;
    s_completed_stages = 0;
    s_start_stage_fade_out_timer = false;
    s_stage_fade_out_timer = 0;
    for (s32 k = 0; k < 10; k++) {
        s_timer_group[k] = {};
        s_completed_stages_world[k] = 0;
    }
}
```

Then in your tick function, just call the function here.
```
if (mkb::scen_info.mode == 5 || mkb::scen_info.mode == 21) {
    // 5 is the file select screen, 21 is the name entry screen
    reset_timer();
}
```

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+        }
+    }
+
+    bool paused_now = *reinterpret_cast<u32*>(0x805BC474) & 8;
+    // for later use, it's useful to record how many stages we've completed
+    // increment the completed stages by 1 during the init
+    // need to check that the game is not paused to ensure the counter only goes up by 1
+    for (s32 k = 0; k < 10; k++) {
+        if (mkb::scen_info.world == k) {
+            if (!paused_now && mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT) {
+                s_completed_stages_world[k] += 1;
+                s_can_lower_stage_counter = true;
+            }
+
+            // if you retry after SMD_GAME_GOAL_INIT but before returning to the stage select
+            // screen, lower the counter by exactly 1
+            if (s_can_lower_stage_counter && mkb::sub_mode == mkb::SMD_GAME_READY_INIT) {
+                s_completed_stages_world[k] += -1;
+                s_can_lower_stage_counter = false;
+            }
+
+            // once you leave a stage, set this to false to ensure the completed stage count is not
+            // lowered when entering spin in on the next stage
+            if (mkb::sub_mode == mkb::SMD_GAME_SCENARIO_RETURN) {
+                s_can_lower_stage_counter = false;
+            }
+        }
+        s_completed_stages += s_completed_stages_world[k];
+    }
```
I think the number of completed stages in the current world is stored in `mkb::mode_info`? Would be worth taking a look at Ghidra because there may be some values that are already being calculated for you that you could use to simplify this.

ComplexPlane replied:
I mentioned before that there's `mkb::scen_info.world` and `mkb::get_world_unbeaten_stage_count()`

eddynya replied:
I simplified the code using `mkb::get_world_unbeaten_stage_count()`

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+                // stop world k's segment timer on tape break of the last stage of the world; we
+                // need to correct by 2f since SMD_GAME_GOAL_INIT happens 2f after tape break
+                s_timer_group[k].segment += -2;
+            }
```
This part can probably be removed if you use the goal entry calculation from my Validation PR as well, but manually subtracting the 2 frames is fine.

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+                // increment the timer every frame during gameplay
+                s_timer_group[k].gameplay += 1;
+            }
+            if (is_postgoal && s_stage_fade_out_timer <= STAGE_FADE_OUT_TIME) {
+                // increment the timer every frame after game goal init happens; once you press
+                // stage select, a separate 49 frame timer is started (fade out from stage
+                // select to the first completely white frame takes 49 frames). once the timer
+                // hits 49 frames, stop incrementing the timer until the 10 ball screen starts
+                // spinning in
+                s_timer_group[k].non_gameplay += 1;
+            }
```
Is there a reason you need to track a timer for this, does using submodes not work?

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+            }
+
+            if (is_on_gameplay) {
+                // increment the timer every frame during gameplay
+                s_timer_group[k].gameplay += 1;
```
Technically the tape-rip frame and the frame after will both be counted for gameplay time, would you want to correct that since it'd be a bonus 2 frames counted per stage?

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+        fullgame_timer_location_y = 2;
+    }
+
+    bool display_story_timer = 0;
```
`bool display_story_timer = false;`

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+        case TimerOptions::EndOfRun:
+            display_story_timer = is_run_complete;
+            break;
+        case TimerOptions::DontShow:
+            display_story_timer = false;
+            break;
+    }
+
+    if (display_story_timer) {
+        timerdisp::draw_timer(FULLGAME_TIMER_LOCATION_X, fullgame_timer_location_y,
+                              FULLGAME_TIMER_TEXT_OFFSET, "Time:", s_loadless_story_timer, 0, false,
+                              false, draw::WHITE);
+    }
+
+    // move the position of the segment timer depending on if the death counter and fullgame timer
+    // is on; if the fullgame timer and death counter is off but the segment timer is on, move the
+    // segment timer to the top line; if either the fullgame timer or death counter are on but not
+    // both are on, move it to the 2nd line, if all 3 are enabled, put it on the 3rd line
+    u32 segment_timer_location_y = 2;
+    if (!display_story_timer && !pref::get(pref::BoolPref::ShowDeathCounter)) {
+        segment_timer_location_y = 2;
+    } else if (!display_story_timer || !pref::get(pref::BoolPref::ShowDeathCounter)) {
+        segment_timer_location_y = 3;
+    } else {
+        segment_timer_location_y = 4;
+    }
```
A simpler way with less constant values:
```
u32 segment_timer_y = 2;
if (display_story_timer) { segment_timer_y++; }
if (pref::get(pref::BoolPref::ShowDeathCounter)) { segment_timer_y++; }
```

ComplexPlane replied:
The `timerdisp` refactor I was proposing would do this as well

eddynya replied:
I just used rehtrop's suggestion, the timerdisp refactor will probably be the last thing I look into once everything else is resolved

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+
+    bool is_run_complete = (mkb::scen_info.world == 9 && s_completed_stages_world[9] == 10);
+
+    // move the positions of the fullgame if the death counter is on
+    u32 fullgame_timer_location_y = 2;
+    if (pref::get(pref::BoolPref::ShowDeathCounter)) {
+        fullgame_timer_location_y = 3;
+    } else {
+        fullgame_timer_location_y = 2;
+    }
```
Simpler code:
```
u32 fullgame_timer_y = 2;
if (pref::get(pref::BoolPref::ShowDeathCounter)) { fullgame_timer_y++; }

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+        segment_timer_location_y = 4;
+    }
+
+    switch (TimerOptions(pref::get(pref::U8Pref::SegmentTimerOptions))) {
```
You should probably include the EndOfRun case here (even if it does nothing)

rehtropsmb commented on src/mods/storytimer.cpp:
```diff
@@ -0,0 +1,368 @@
+            break;
+    }
+
+    // if the segment timer is enabled in any capacity, show all 10 split times + iw times after the
+    // tape is broken on the last stage
+    if (TimerOptions(pref::get(pref::U8Pref::SegmentTimerOptions)) != TimerOptions::DontShow &&
+        is_run_complete) {
+        // haven't attempted the mkb::sprintf suggestion yet
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y, IW_TIME_TEXT_OFFSET,
+                              "W1:", s_timer_group[0].split, s_timer_group[0].segment, true, false,
+                              draw::WHITE);
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y + 1, IW_TIME_TEXT_OFFSET,
+                              "W2:", s_timer_group[1].split, s_timer_group[1].segment, true, false,
+                              draw::WHITE);
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y + 2, IW_TIME_TEXT_OFFSET,
+                              "W3:", s_timer_group[2].split, s_timer_group[2].segment, true, false,
+                              draw::WHITE);
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y + 3, IW_TIME_TEXT_OFFSET,
+                              "W4:", s_timer_group[3].split, s_timer_group[3].segment, true, false,
+                              draw::WHITE);
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y + 4, IW_TIME_TEXT_OFFSET,
+                              "W5:", s_timer_group[4].split, s_timer_group[4].segment, true, false,
+                              draw::WHITE);
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y + 5, IW_TIME_TEXT_OFFSET,
+                              "W6:", s_timer_group[5].split, s_timer_group[5].segment, true, false,
+                              draw::WHITE);
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y + 6, IW_TIME_TEXT_OFFSET,
+                              "W7:", s_timer_group[6].split, s_timer_group[6].segment, true, false,
+                              draw::WHITE);
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y + 7, IW_TIME_TEXT_OFFSET,
+                              "W8:", s_timer_group[7].split, s_timer_group[7].segment, true, false,
+                              draw::WHITE);
+        timerdisp::draw_timer(IW_TIME_LOCATION_X, segment_timer_location_y + 8, IW_TIME_TEXT_OFFSET,
+                              "W9:", s_timer_group[8].split, s_timer_group[8].segment, true, false,
+                              draw::WHITE);
+        // use segment timer spacing for w10 since "W10" is 3 characters long, not 2
+        timerdisp::draw_timer(SEGMENT_TIMER_LOCATION_X, segment_timer_location_y + 9,
+                              SEGMENT_TIMER_TEXT_OFFSET, "W10:", s_timer_group[9].split,
+                              s_timer_group[9].segment, true, false, draw::WHITE);
+    }
```
Should this be it's own bool preference? Or just the EndOfRun case?

eddynya replied:
The way it works now is that if the segment timer is on at all (so either showing at all times, between worlds, or at the end of the run), when the run is finished, all 10 split and segment times are shown. I don't think this should get its own bool pref since this information is shown after the run, and so people who want to run pace blind aren't affected by it

rehtropsmb reviewed (changes requested):
Some more stuff to take a look at. I still haven't tested in-game, this was just from doing an overview of the code. I'll do some gameplay testing eventually :)

Also make sure to remove the script changes from your PR! And pull in the most up-to-date code so you can fix your merge conflicts!

Commit:
pre merge commit

eddynya committed:
Update pref.cpp

eddynya committed:
Merge branch 'master' of github.com:ComplexPlane/SMB2PracticeMod

eddynya committed:
fixed deathcounter bug

retrying on the tape break frame no longer increases the death counter

eddynya committed:
started addressing storytimer comments

nothing significant really

eddynya committed:
small changes before bigger changes

eddynya committed:
backup before fixing bugs

-timer and seg timer are consistently offset from each other currently when they shouldn't be

eddynya committed:
removed unnecessary static variables in the TimerGroup struct

also fixed the update order bug causing the segment timer to be 1f behind what it should be

eddynya committed:
simplified code for the number of completed stages

eddynya committed:
removed the s_completed_stages_world array

eddynya committed:
Delete storytimer.cppZone.Identifier

eddynya committed:
removed unused variables

eddynya committed:
rewrote end of run segment timer code using mkb::sprintf

eddynya committed:
fixed some end of run segment formatting

issue: the w1 split does not show up

eddynya committed:
fixed the w1 not showing issue

eddynya committed:
improved code for stopping the segment timer

the segment timer now stops on the goal tape break frame instead of using SMD_GOAL_INIT and subtracting 2 frames manually

eddynya committed:
clean up

eddynya committed:
reverted draw_timer function to only draw one timer

eddynya committed:
debugging stuff

eddynya committed:
removed debugging variables/cleanup

eddynya committed:
Merge branch 'ComplexPlane:master' into master

rehtropsmb committed:
update scripts

eddynya committed:
minor things

-removed unused variables
-fixed error where split 1 was off by 1f

eddynya committed:
Merge branch 'master' of https://github.com/eddynya/SMB2PracticeMod

ComplexPlane commented:
Closing, new PR is #88
